Fix: Attemp to retrieve traffic signal information of nodes through trace_attribute request

MohamedM216 opened the pull request:
# Issue

fixes #5106: **Retrieve traffic signal information of nodes through trace_attribute request**.

## Tasklist

- [x] Attempt to add support for `node.traffic_signal` attribute in `/trace_attributes` API.
- [x] Add tests to verify the new functionality (pending guidance).
- [x] Update the docs with any new request parameters or changes to behavior described (pending guidance).
- [x] Update the [changelog](CHANGELOG.md) (pending guidance).
- [ ] If you made changes to the lua files, update the [taginfo](taginfo.json) too.

## Requirements / Relations

This PR is based on the discussion in issue #5106. No other pull requests are directly related to this change.

---

### Description
This PR is an attempt to add support for retrieving the `node.traffic_signal` attribute through the `/trace_attributes` API. The `traffic_signal` boolean in `baldr/nodeinfo.h` is intended to be exposed in the API response, allowing users to query traffic signal information for nodes along a route.

However, after making the changes, the `traffic_signal` attribute is not appearing in the API response. I’m seeking guidance on how to proceed to resolve this issue.

### Changes
1. **Added `kNodeTrafficSignal` to `AttributesController::kDefaultAttributes`**:
   - Enabled the `node.traffic_signal` attribute by default in the `AttributesController`.

2. **Modified `serialize_edges` in `trace_serializer.cc`**:
   - Added logic to include the `traffic_signal` attribute in the API response when requested.

3. **Defined `kNodeTrafficSignal` in `attributes_controller.h`**:
   - Added the `node.traffic_signal` key to the node attributes.

### Testing
- Tested the changes using the Andorra OSM dataset (`andorra-latest.osm.pbf`).
- The `node.traffic_signal` attribute is **not appearing** in the `/trace_attributes` response, even though edge attributes like `edge.speed` and `edge.length` are returned.

Example request:

```bash
curl -X POST http://localhost:8002/trace_attributes \
  -H "Content-Type: application/json" \
  -d '{
    "shape": [
      {"lat": 42.5063, "lon": 1.5218},
      {"lat": 42.5075, "lon": 1.5244}
    ],
    "costing": "auto",
    "shape_match": "map_snap",
    "filters": {
      "attributes": [
        "node.traffic_signal",
        "edge.speed",
        "edge.length"
      ],
      "action": "include"
    },
    "trace_options": {
      "search_radius": 50,
      "gps_accuracy": 5.0
    }
  }' | jq
```

And here’s the response I received:

```json
    {
  "units": "kilometers",
  "edges": [
    {
      "speed": 50,
      "length": 0.057
    },
    {
      "speed": 50,
      "length": 0.044
    },
    {
      "speed": 50,
      "length": 0.024
    },
    {
      "speed": 50,
      "length": 0.021
    },
    {
      "speed": 50,
      "length": 0.047
    },
    {
      "speed": 50,
      "length": 0.015
    },
    {
      "speed": 50,
      "length": 0.02
    },
    {
      "speed": 25,
      "length": 0.006
    },
    {
      "speed": 25,
      "length": 0.029
    }
  ],
  "alternate_paths": []
}
```

kevinkreiser commented:
adhoc testing (running a server and sending a curl request to check the result) is a good start but to merge this and to make it easier to debug if there is a problem we want to make some unit/integration tests. the good (or maybe bad depending on your perspective haha) news is that we have tons of these already that you can model a new test off of.

specifically you should add to this suite of tests: https://github.com/valhalla/valhalla/blob/master/test/gurka/test_trace_attributes.cc

there is a whole directory of "gurka" examples. you should check out this readme: https://github.com/valhalla/valhalla/blob/master/test/gurka/README.md

the structure allows you to make a road network from fake data, create valhalla tiles from it and make sure that the library does what is expected based on that fake data

MohamedM216 commented:
Hi @kevinkreiser,

Thank you for your feedback! I've updated the PR to include a unit/integration test in the `gurka` suite, as you suggested. Specifically, I added a new test case in `test_trace_attributes.cc` to verify the `traffic_signal` functionality.

### Here’s a summary of the changes:
- Added a new test case in `test_trace_attributes.cc` to check the `traffic_signal` attribute in the `trace_attributes` response.

- The `ascii_map` I used is very simple. After building tiles, I send an API request with only 2 edges: `AB` and `BC`. Edge `AB`  should have `traffic_signal : true` and edge `BC` should have `traffic_signal : false`. The response has these two attributes with the correct values.

- The test passed locally and is consistent with the existing test suite.

- The test now checks for the `traffic_signal` field in the `end_node` of the `edges`   array, as this is where the attribute is located in the response.

### API Response:
```json
{
  "units": "kilometers",
  "osm_changeset": 1001,
  "shape": "???mU?oU",
  "confidence_score": 1,
  "raw_score": 0,
  "admins": [
    {
      "country_text": "None",
      "state_text": "None"
    }
  ],
  "edges": [
    {
      "truck_route": false,
      "density": 0,
      "sac_scale": 0,
      "shoulder": false,
      "sidewalk": "none",
      "bicycle_network": 0,
      "cycle_lane": "none",
      "lane_count": 1,
      "max_downward_grade": -500,
      "max_upward_grade": -500,
      "weighted_grade": 0,
      "way_id": 4,
      "id": 33570632,
      "travel_mode": "drive",
      "vehicle_type": "car",
      "surface": "paved_smooth",
      "drive_on_right": false,
      "internal_intersection": false,
      "roundabout": false,
      "bridge": false,
      "tunnel": false,
      "unpaved": false,
      "toll": false,
      "use": "road",
      "traversability": "both",
      "end_shape_index": 1,
      "begin_shape_index": 0,
      "end_heading": 90,
      "begin_heading": 90,
      "road_class": "primary",
      "speed": 75,
      "country_crossing": false,
      "forward": true,
      "length": 0.04,
      "names": [
        "AB"
      ],
      "end_node": {
        "intersecting_edges": [
          {
            "walkability": "both",
            "cyclability": "both",
            "driveability": "both",
            "from_edge_name_consistency": false,
            "to_edge_name_consistency": false,
            "begin_heading": 180,
            "use": "road",
            "road_class": "primary"
          }
        ],
        "elapsed_time": 1.92,
        "elapsed_cost": 1.632,
        "admin_index": 0,
        "type": "street_intersection",
        **"traffic_signal": true,**
        "fork": false,
        "transition_time": 2.017
      }
    },
    {
      "truck_route": false,
      "density": 0,
      "sac_scale": 0,
      "shoulder": false,
      "sidewalk": "none",
      "bicycle_network": 0,
      "cycle_lane": "none",
      "lane_count": 1,
      "max_downward_grade": -500,
      "max_upward_grade": -500,
      "weighted_grade": 0,
      "way_id": 5,
      "id": 100679496,
      "travel_mode": "drive",
      "vehicle_type": "car",
      "surface": "paved_smooth",
      "drive_on_right": false,
      "internal_intersection": false,
      "roundabout": false,
      "bridge": false,
      "tunnel": false,
      "unpaved": false,
      "toll": false,
      "use": "road",
      "traversability": "both",
      "end_shape_index": 2,
      "begin_shape_index": 1,
      "end_heading": 90,
      "begin_heading": 90,
      "road_class": "primary",
      "speed": 75,
      "country_crossing": false,
      "forward": true,
      "length": 0.04,
      "names": [
        "BC"
      ],
      "end_node": {
        "elapsed_time": 5.857,
        "elapsed_cost": 9.763,
        "admin_index": 0,
        "type": "street_intersection",
        **"traffic_signal": false,**
        "fork": false,
        "transition_time": 0
      }
    }
  ],
  "matched_points": [
    {
      "lon": 0,
      "lat": 0,
      "type": "matched",
      "edge_index": 0,
      "distance_along_edge": 0,
      "distance_from_trace_point": 0
    },
    {
      "lon": 0.000359,
      "lat": 0,
      "type": "matched",
      "edge_index": 0,
      "distance_along_edge": 1,
      "distance_from_trace_point": 0.035914
    },
    {
      "lon": 0.000719,
      "lat": 0,
      "type": "matched",
      "edge_index": 1,
      "distance_along_edge": 1,
      "distance_from_trace_point": 0.038738
    }
  ],
  "alternate_paths": []
}
```

Please let me know if there’s anything else I need to address or if there are further improvements I can make.

Thanks again for your guidance!

kevinkreiser commented on CHANGELOG.md:
```diff
@@ -1,6 +1,7 @@
 ## UNRELEASED
 * **Removed**
 * **Bug Fix**
+   * FIXED: Attemp to retrieve traffic signal information of nodes through trace_attribute request [#5121](https://github.com/valhalla/valhalla/pull/5121)
```
its not a bug fix, its something new we are adding. also our changelog has the newer stuff at the bottom of each section rather than the top

MohamedM216 replied:
For the first update I think we don't need it as an `edge` may include `end_node` which includes `traffic_signal`.

MohamedM216 replied:
I've updated the changelog and the Map Matching API documentation (`docs/api/map-matching/api-reference.md`). The changes have been committed and pushed, and the PR has been updated. Please, let me know if there is any thing else I need to address before merging.

If possible, I'd love to take on another issue from the list of chosen issues you shared earlier here: https://github.com/valhalla/valhalla/issues/5099. Could you recommend one that would be a good next step for me?

Looking forward to your feedback!
